### ISSUE TITLE
Cannot Show-UDToast from within an Endpoint

### ISSUE THREAD

**Issue by mrhockeymonkey** [ironmansoftware/universal-dashboard#706]
#439 suggests this issue was fixed/implemented but I haven't been able to get this working. Here is my code:

```powershell
New-UDEndpoint -Url "/doSomething/:id" -Method "GET" -Id 'dosomething' -Endpoint {
    Param ($Id)
    Show-UDToast -Message "Something has been done" -Title 'Done' -Position topRight
    "done"
}
```

Is this expected to work? Interestingly also giving this endpoint an id doesn't make it show up in developer mode in chrome. But when hitting it with invoke-restmethod the string "done" is returned so it is working.

**Comment by adamdriscoll** [comment]
Yeah that should work. Can you grab a log? `Enable-UDLogging`

**Comment by mrhockeymonkey** [comment]
So here are the logs for the point in time I call the api. Notably there is no SessionId or ConnectionId but the Error level log just says one of more errors occurred

```plain
12:01:48 [Info] Microsoft.AspNetCore.Hosting.Internal.WebHost Request starting HTTP/1.1 GET http://whzwindh1:5000/api/doSomething/hello
12:01:48 [Info] Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker Route matched with {action = "GetEndpoint", controller = "Component"}. Executing action UniversalDashboard.Controllers.ComponentController.GetEndpoint (UniversalDashboard)
12:01:48 [Info] Microsoft.AspNetCore.Authorization.DefaultAuthorizationService Authorization was successful.
12:01:48 [Debug] ExecutionService ExecutionService constructor
12:01:48 [Info] Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker Executing action method UniversalDashboard.Controllers.ComponentController.GetEndpoint (UniversalDashboard) - Validation state: Valid
12:01:48 [Info] ComponentController GetEndpoint - doSomething/hello
12:01:48 [Debug] ExecutionService Context variables--------------
12:01:48 [Debug] ExecutionService Request = Microsoft.AspNetCore.Http.Internal.DefaultHttpRequest
12:01:48 [Debug] ExecutionService Response = Microsoft.AspNetCore.Http.Internal.DefaultHttpResponse
12:01:48 [Debug] ExecutionService User =
12:01:48 [Debug] ExecutionService MemoryCache = Microsoft.Extensions.Caching.Memory.MemoryCache
12:01:48 [Debug] ExecutionService DashboardHub = Microsoft.AspNetCore.SignalR.Internal.HubContext`1[UniversalDashboard.DashboardHub]
12:01:48 [Debug] ExecutionService Cache = Microsoft.Extensions.Caching.Memory.MemoryCache
12:01:48 [Debug] ExecutionService StateRequestService = UniversalDashboard.Services.StateRequestService
12:01:48 [Debug] ExecutionService ConnectionId =
12:01:48 [Debug] ExecutionService SessionId =
12:01:48 [Debug] ExecutionService ArgumentList =
12:01:48 [Debug] ExecutionService UDPage =
12:01:48 [Debug] ExecutionService ClaimsPrinciple = System.Security.Claims.ClaimsPrincipal
12:01:48 [Debug] ExecutionService ClaimsPrincipal = System.Security.Claims.ClaimsPrincipal
12:01:48 [Debug] DashboardColor DashboardColor - color = Color [A=153, R=0, G=0, B=0]
12:01:49 [Debug] ColorExtensions rgba(0, 0, 0, 0.6)
12:01:49 [Error] ShowToastCommand One or more errors occurred.
12:01:49 [Debug] ComponentController ConvertToActionResult() done (hello)
12:01:49 [Info] Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker Executed action method UniversalDashboard.Controllers.ComponentController.GetEndpoint (UniversalDashboard), returned result Microsoft.AspNetCore.Mvc.ContentResult in 384.0709ms.
12:01:49 [Info] Microsoft.AspNetCore.Mvc.Infrastructure.ContentResultExecutor Executing ContentResult with HTTP Response ContentType of application/json; charset=utf-8
12:01:49 [Info] Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker Executed action UniversalDashboard.Controllers.ComponentController.GetEndpoint (UniversalDashboard) in 486.44ms
12:01:49 [Info] Microsoft.AspNetCore.Hosting.Internal.WebHost Request finished in 670.7222ms 200 application/json; charset=utf-8
```

And for comparison here are the logs for when I use Show-UDToast from a button -OnClick method. 

```plain
12:05:59 [Debug] ExecutionService ExecutionService constructor
12:05:59 [Debug] DashboardHub ClientEvent dc67d743-9f93-4b80-aa0d-b4024a63ce11onClick onClick
12:05:59 [Debug] EndpointService Get() dc67d743-9f93-4b80-aa0d-b4024a63ce11onClick dad70f6d-77a2-4b54-a708-5272e4b16e24
12:05:59 [Debug] EndpointService Found session endpoint.
12:05:59 [Debug] ExecutionService Scope variables--------------
12:05:59 [Debug] ExecutionService Context variables--------------
12:05:59 [Debug] ExecutionService EventData =
12:05:59 [Debug] ExecutionService MemoryCache = Microsoft.Extensions.Caching.Memory.MemoryCache
12:05:59 [Debug] ExecutionService DashboardHub = Microsoft.AspNetCore.SignalR.Internal.HubContext`1[UniversalDashboard.DashboardHub]
12:05:59 [Debug] ExecutionService Cache = Microsoft.Extensions.Caching.Memory.MemoryCache
12:05:59 [Debug] ExecutionService StateRequestService = UniversalDashboard.Services.StateRequestService
12:05:59 [Debug] ExecutionService ConnectionId = 0JsdDcnsRbUaP0586zcsTQ
12:05:59 [Debug] ExecutionService SessionId = dad70f6d-77a2-4b54-a708-5272e4b16e24
12:05:59 [Debug] ExecutionService ArgumentList =
12:05:59 [Debug] ExecutionService UDPage =
12:05:59 [Debug] ExecutionService ClaimsPrinciple = System.Security.Claims.ClaimsPrincipal
12:05:59 [Debug] ExecutionService ClaimsPrincipal = System.Security.Claims.ClaimsPrincipal
12:05:59 [Debug] DashboardColor DashboardColor - color = Color [A=153, R=0, G=0, B=0]
12:05:59 [Debug] ColorExtensions rgba(0, 0, 0, 0.6)
12:06:14 [Debug] Quartz.Core.QuartzSchedulerThread Batch acquisition of 0 triggers
```

Its worth noting also that after some time has passed the button also seems to break and some errors pop up on the screen (not logged) that read "Websocket closed with status code: 1006()". Not sure if this is related.

Let me know if there is any other steps i can take to try and highlight the issue further

**Comment by mrhockeymonkey** [comment]
In case it helps here are the initial startup logs also. Albeit from a point later in time but of the same dashboard. 

```plain
12:14:33 [Info] StartDashboardCommand  - C:\DSC.Dash\DashboardTest.ps1 - True
12:14:33 [Debug] StartDashboardCommand {"id":"00000000-0000-0000-0000-000000000000","title":"DSC Dashboard","navBarColor":null,"navBarFontColor":null,"backgroundColor":null,"fontColor":null,"fontIconStyle":"FontAwesome","navbarLinks":null,"scripts":null,"stylesheets":null,"pages":[{"type":"page","name":"Home","url":null,"defaultHomePage"
:false,"icon":"home","components":[{"tag":"div","attributes":{"className":"row"},"events":[],"content":[{"tag":"div","attributes":{"className":"col s6"},"events":[],"type":"element","key":"b5aef191-dcff-4829-a962-125a82e05d65","onMount":null,"id":"a9e2e6f8-c83d-4b8b-982c-097da2f9d8f2","refreshInterval":5,"autoRefresh":false,"hasCallback"
:true}],"type":"element","key":"439b48f5-069b-43c1-a67c-1a4c29264f58","onMount":null,"id":"f0cc86db-2f20-4080-b066-d05a36b7e72f","refreshInterval":5,"autoRefresh":false,"hasCallback":false}],"dynamic":false,"id":"2b6645c3-25fd-4126-96e4-32aa31614278","refreshInterval":5,"autoRefresh":false,"hasCallback":false}],"cyclePages":false,"cycleP
agesInterval":10,"error":null,"design":false,"footer":null,"navBarLogo":null,"loginPage":null,"Demo":false,"geolocation":false,"filterText":null,"navigation":null,"loadingScreen":null}
12:14:33 [Debug] CustomAssemblyLoadContext Pointer Size:8
12:14:33 [Debug] CustomAssemblyLoadContext Loading files in native directory: C:\Program Files\WindowsPowerShell\Modules\UniversalDashboard\2.3.2\net472\runtimes\win-x64\native
12:14:33 [Debug] CustomAssemblyLoadContext Loading file: C:\Program Files\WindowsPowerShell\Modules\UniversalDashboard\2.3.2\net472\runtimes\win-x64\native\libuv.dll
12:14:33 [Debug] CustomAssemblyLoadContext Loading file: C:\Program Files\WindowsPowerShell\Modules\UniversalDashboard\2.3.2\net472\runtimes\win-x64\native\Microsoft.DiaSymReader.Native.amd64.dll
12:14:33 [Debug] EndpointService Register() dosomething /doSomething/:id
12:14:34 [Debug] ExecutionService ExecutionService constructor
12:14:35 [Info] Quartz.Impl.StdSchedulerFactory Using object serializer: Quartz.Simpl.BinaryObjectSerializer, Quartz
12:14:35 [Debug] Quartz.Simpl.TaskSchedulingThreadPool TaskSchedulingThreadPool configured with max concurrency of 10 and TaskScheduler ThreadPoolTaskScheduler.
12:14:35 [Info] Quartz.Core.SchedulerSignalerImpl Initialized Scheduler Signaller of type: Quartz.Core.SchedulerSignalerImpl
12:14:35 [Info] Quartz.Core.QuartzScheduler Quartz Scheduler v.3.0.6.0 created.
12:14:35 [Info] Quartz.Simpl.RAMJobStore RAMJobStore initialized.
12:14:35 [Info] Quartz.Core.QuartzScheduler Scheduler meta-data: Quartz Scheduler (v3.0.6.0) 'QuartzScheduler' with instanceId 'NON_CLUSTERED'
  Scheduler class: 'Quartz.Core.QuartzScheduler' - running locally.
  NOT STARTED.
  Currently in standby mode.
  Number of jobs executed: 0
  Using thread pool 'Quartz.Simpl.DefaultThreadPool' - with 10 threads.
  Using job-store 'Quartz.Simpl.RAMJobStore' - which does not support persistence. and is not clustered.

12:14:35 [Info] Quartz.Impl.StdSchedulerFactory Quartz scheduler 'QuartzScheduler' initialized
12:14:35 [Info] Quartz.Impl.StdSchedulerFactory Quartz scheduler version: 3.0.6.0
12:14:35 [Info] Quartz.Core.QuartzScheduler Scheduler QuartzScheduler_$_NON_CLUSTERED started.
Hosting environment: Production
Content root path: C:\Program Files\WindowsPowerShell\Modules\UniversalDashboard\2.3.2\net472\..\client
Now listening on: http://0.0.0.0:5000
Application started. Press Ctrl+C to shut down.
12:14:35 [Debug] Quartz.Core.QuartzSchedulerThread Batch acquisition of 0 triggers
```

Here is the full dashboard (apologies there is some unused junk in there I neglected to clean out but didnt think it would affect this test dashboard)

```powershell
$Colours = @{
    Red = '#f70e1e'
    Green = '#04ce5b'
    Blue = '#2954f2'
    DarkBlue = '#01229b'
    Grey = '#919191'
    Blue1 = '#62aac3'
    Blue2 = '#318fbb'
    Blue3 = '#2c6e91'
    Blue4 = '#20364b'
    Blue5 = '#171738'
}

$BluePallete = @(
    $Colours['Blue1']
    $Colours['Blue2']
    $Colours['Blue3']
    $Colours['Blue4']
    $Colours['Blue5']
)

$Index = . (Join-Path $PSScriptRoot "pages\Index.ps1")
#$Reports = . (Join-Path $PSScriptRoot "pages\Reports.ps1")
#$NodeRSOC = . (Join-Path $PSScriptRoot "pages\NodeRSOC.ps1")
#$Errors = . (Join-Path $PSScriptRoot "pages\Errors.ps1")

$Every30Seconds = New-UDEndpointSchedule -Every 30 -Second
$EveryMinute = New-UDEndpointSchedule -Every 1 -Minute
$Every30Minutes = New-UDEndpointSchedule -Every 30 -Minute



$DS = New-UDEndpoint -Url "/doSomething/:id" -Method "GET" -Id 'dosomething' -Endpoint {
    Param ($Id)
    Show-UDToast -Message "Something has been done ($Id)" -Title 'Done' -Position topRight
    "done ($Id)"
}

# Start Dashboard
Get-UDDashboard | Stop-UDDashboard
$Dashboard = New-UDDashboard -Title "DSC Dashboard" -Pages @(
    New-UDPage -Name "Home" -Icon home -Content {
        New-UDRow -Columns {
            New-UDColumn -SmallSize 6 -Endpoint {
                New-UDButton -Text press -OnClick {
                    Show-UDToast -Message "Button Pressed" -Title 'Pressed' -Position topRight
                }
            }
        }
    }
)
Enable-UDLogging -Level Debug
Start-UDDashboard -Dashboard $Dashboard -Port 5000 -AutoReload -Wait -Endpoint @(
    $DS
)

```

**Comment by adamdriscoll** [comment]
I totally spaced on this. This configuration will not work. I was looking at this as if the New-UDEndpoint was being used with a dashboard control. New-UDEndpoint REST API won't show a toast because it doesn't have a session. 

What could be added would be a `-Broadcast` parameter that would then show a toast on all connected dashboards. Would that implement what you are looking for?

**Comment by adamdriscoll** [comment]
So I ended up adding a -Broadcast parameter that will make this would. Will be available in 2.4-beta2.

**Comment by mrhockeymonkey** [comment]
Yeah that works for me. Ill keep an eye out for this release 👍
